src/slicer.cpp

```diff
@@ -82,122 +84,305 @@ int SlicerLayer::getNextSegmentIdx(const Mesh* mesh, const SlicerSegment& segmen
+        if (completed_poly)
+        {
+            // finished polygon
+            PolygonRef polyline_0 = open_polylines[best_polyline_0_idx];
+            polygons.add(polyline_0);
+            polyline_0.clear();
+            unsigned int cur_terms[2] = {best_polyline_0_idx*2 + 0,
+                                         best_polyline_0_idx*2 + 1};
+            unsigned int old_terms[2];
+            for (size_t idx=0U; idx!=2U; ++idx)
             {
-                if (polyline_1.size() > polyline_2.size()) // decide which polygon to copy into the other
-                {
-                    for(int poly_idx = polyline_2.size()-1; poly_idx >= 0; poly_idx--)
-                        polyline_1.add(polyline_2[poly_idx]);
-                    polyline_2.clear();
-                } 
-                else
+                old_terms[idx] = terminus_cur_to_old_map[cur_terms[idx]];
+            }
+            for (size_t idx=0U; idx!=2U; ++idx)
+            {
+                terminus_old_to_cur_map[old_terms[idx]] = INVALID_TERMINUS;
+                terminus_cur_to_old_map[cur_terms[idx]] = INVALID_TERMINUS;
+            }
+            continue;
+        }
+
+        // plan how to append polygons
+        bool back_0 = (terminus_0_idx & 1) == 1;
+        bool back_1 = (terminus_1_idx & 1) == 1;
+        bool reverse[2] = {false, false};
+        if (back_0)
+        {
+            if (back_1)
+            {
+                if (open_polylines[best_polyline_0_idx].size() <
+                    open_polylines[best_polyline_1_idx].size())
                 {
-                    for(int poly_idx = polyline_1.size()-1; poly_idx >= 0; poly_idx--)
-                        polyline_2.add(polyline_1[poly_idx]);
-                    polyline_1.clear();
+                   std::swap(terminus_0_idx,terminus_1_idx);
                 }
-                // note that either way we end up with the end of former polyline_1 next to the start of former polyline_2
+                reverse[1] = true;
+            } else {
+                // nothing to do
+            }
+        } else {
+            if (back_1)
+            {
+                std::swap(terminus_0_idx,terminus_1_idx);
             }
             else
             {
-                for(Point& p : polyline_2)
-                    polyline_1.add(p);
-                polyline_2.clear();
+                reverse[0] = true;
+            }
+        }
+
+        best_polyline_0_idx = terminus_0_idx/2;
+        best_polyline_1_idx = terminus_1_idx/2;
+        PolygonRef polyline_0 = open_polylines[best_polyline_0_idx];
+        PolygonRef polyline_1 = open_polylines[best_polyline_1_idx];
+
+        // append polygons according to plan
+        if (reverse[0])
+        {
+            // reverse polyline_0
+            size_t size_0 = polyline_0.size();
+            for (size_t idx=0U; idx!=size_0/2; ++idx)
+            {
+                std::swap(polyline_0[idx], polyline_0[size_0-1-idx]);
```

Comment:
It seems to me that this reordering is unnecessary if `reverse[1]`.

We could keep the ordering reversed and simply append `polyline_0` to `polyline_1` and clear `polyline_0`.

What do you think?